### ISSUE TITLE
Particles Animation repeat the first frame when using not square sheets

### ISSUE THREAD

**Issue by Anyeos** [godotengine/godot#36435]
**Godot version:**
3.2 stable

**OS/device including version:**
Ubuntu 19.10

**Issue description:**
The Particles2D with Particles Animation repeat the first frame when it animates.
On a 3x2 sheet the Particles Animation shows in this order: 1, 2, 3, 1, 5, 6. But it must be: 1, 2, 3, 4, 5, 6
The sheet is correctly created from left to right and uppter to bottom.
I numerated the sheet and it repeat the number "1". So it is eventualy wrong.

**Steps to reproduce:**
1) Add a Particles2D node.
2) Load a texture with a sheet of animation (ie: Explosion)
3) Create a CanvasItemMaterial and set it to animate and fill the H and V frames properly.

**Minimal reproduction project:**

**Comment by Anyeos** [comment]
![muzzle1](https://user-images.githubusercontent.com/5976100/75085273-44c2c000-5506-11ea-947f-e22dbfe28871.png)

![Simple0000-2020-02-22_00 06 48](https://user-images.githubusercontent.com/5976100/75085394-696b6780-5507-11ea-9b6b-78943a0c26bc.gif)

**Comment by Anyeos** [comment]
Workaround: Instead of 3x2 sheet I created a 2x3 sheet with the same frames and it works as expected.

**Comment by tom-jk** [comment]
Can't reproduce in 3.2.2.stable.

Minimum reproduction project: [36435.zip](https://github.com/godotengine/godot/files/4908780/36435.zip)

[Video (0m10s)](https://youtu.be/C_QuRetaeag)

**Comment by Calinou** [comment]
@tom-jk Thanks for testing it! I'll close this as resolved then.

If anyone else can still reproduce this in 3.2.2 or later, please comment so this issue can be reopened.

**Comment by Anyeos** [comment]
The problem is still present but different behaviour. And the same behaviour in CPUParticles2D. So it appears to be a bug inside of the sheet handling.


[Bug#36435.zip](https://github.com/godotengine/godot/files/5314519/Bug.36435.zip)

**Comment by metinc** [comment]
I found another discussion on this topic where the conclusion was that it must be related to older hardware or drivers: #31699

I can reproduce this bug on v3.3.3.stable.official [b973f997f]. I'm using Ubuntu 21.04 but my hardware is not old (AMD Ryzen 7 4800H with Radeon Graphics).

**Comment by metinc** [comment]
I looked into this issue and found out that the issue is caused by an imprecision in the shader when calling floor(). For example the parameter in floor() could be expected to be 1.0 and should be floored to 1.0 but in fact it is something like 0.999999 and gets floored to 0.0 (because of the imprecision). Have a look at my bugfix for the exact line of code #53233 . You will see that I added a 0.5 to the number of frames before using floor which ensures we wont get close to number where imprecision might happen.
